tests/regression/13-privatized/67-pthread_cond_wait.c

```diff
@@ -12,6 +12,7 @@ void* f1(void* ptr) {
     pthread_mutex_lock(&mut);
     g = 1;
     pthread_cond_wait(&cond,&mut);
+    // TODO: mutex-oplus is unsound here!
```

Comment:
Just an observation from running the tests with mutex-oplus as default. For comparison I also ran it before my changes and it was already there, so at least I didn't introduce it here, but it was worth noting down somewhere. I think I'll also add a mention to mutex-oplus implementation so it's a bit more visible than this random test, but I don't know how much we care about mutex-oplus, so I didn't want to spend time debugging and fixing it.

Reply:
It seems the problem is the check for `is_unprotected ask x` as that refers to the successor, where the global still was protected. I changed it to `is_unprotected_without ask x m` at that seems to help.

Reply:
That's very weird because indeed all queries run on the state before the transfer function, but this is `lock`, so `m` should be unlocked there anyway and using `_without` which just removes it should make no difference. The `_without` version was added to be able to do the query correctly in some `unlock` to access what the protection information would be after the lockset decreased.

But I think I know why this weird behavior is happening now: `pthread_cond_wait` is implemented in the mutex analysis as two events: unlock and lock. In the base analysis the `event` transfer function gets two contexts `ctx` and `octx`. When events are applied, `ctx` changes as the local state changes between events, but `octx` is the original context before any of the events have been applied. And for some reason I've decided to pass `octx` instead of `ctx` to all the privatizations. This is why `lock` is seeing things already locked: it's seeing the local state before the `unlock` even.
I'm now thinking maybe it should be passing `ctx` instead of `octx` to privatizations...

Reply:
Yes, that indeed seems to be the more systematic approach.

Reply:
Wait, no, git blaming my own changes I know why `octx` is used there. `octx` isn't just the context before all events but after the transfer function (that emitted the events), but it's the same `ctx` that the transfer function used, so the context before the transfer function itself.

This was necessary because all the `lock`/`unlock` implementations are such that they assume querying the state before the lock/unlock statement, where the mutex queries come out right. Because it's the mutex analysis that both changes the current lockset and emits the events, all the `lock`/`unlock` would have to be rewritten to work with the post-locksets of those operations instead. It would be useful for the `is_unprotected_without` usages, but it would be far worse for the `is_unprotected` usages, because they would need special handling instead where the just-added lock is excluded.

For this to properly work the event system would somehow have to be even more involved because the events somehow need to be able to choose which `ctx` they want to operate on and how they should be chained-grouped. And it's not really possible to have one analysis trigger an event that should run before an another analysis runs the corresponding transfer function. All this event interactions matter in these subtle ways.

So I think the systematic approach isn't as simple as I initially expected and I guess we'll have to go with this hack for the time being.

Reply:
I now started worrying that maybe other privatizations could have the same issue, but it seems not, because they don't use `is_unprotected`, but only `is_protected_by`, which isn't a local query but a global query.

I'll add some kind of comment about this insanity and merge this.